### PR TITLE
Split convolution invocation into preparation and actual invocation

### PR DESCRIPTION
**Pull request by whchung**
- split DoConvolve into:

PrepareForConvolution
DoConvolve

- split DoConvolveBackwardData into:

PrepareForConvolutionBackwardData
DoConvolveBackwardData

- split DoConvolveBackwardFilter into:

PrepareForConvolutionBackwardFilter
DoConvolveBackwardFilter

PrepareForConvolutionXXX would allocate scratch memory.
DoConolveXXX would invoke actual convolution algorithms.

Implement forward convoution, backward input convolution, backward filter
convolution on CUDA path.

### CODE REVIEW

**Comment by whchung** [comment]
@timshen91 per discussion please find my PR to separate convolution invocation into 2 parts here. This PR only covers logic for CUDA path.

Implementation for ROCm shall be submitted after other PRs from @deven-amd for StreamExecutor.

File: tensorflow/stream_executor/dnn.h [hunk 1]
```diff
@@ -730,6 +730,7 @@ class PoolingDescriptor {
 class AlgorithmDesc {
  public:
   typedef int64 Index;
+  AlgorithmDesc() {}
```
Comment (timshen91) [inline tensorflow/stream_executor/dnn.h]: `AlgorithmDesc() : AlgorithmDesc(0, false) {}`

File: tensorflow/stream_executor/stream.cc [hunk 2]
```diff
@@ -758,12 +768,22 @@ Stream &Stream::ThenConvolveWithAlgorithm(
...
-      auto status = dnn->DoConvolve(
+      DeviceMemory<uint8> scratch_memory;
+      dnn::AlgorithmDesc algorithm_desc;
+      auto status = dnn->PrepareForConvolution(
           this, input_descriptor, input_data, filter_descriptor, filter_data,
           convolution_descriptor, output_descriptor, output, scratch_allocator,
-          algorithm_config, output_profile_result);
-      if (!status && !output_profile_result) {
+          algorithm_config, &algorithm_desc, &scratch_memory);
+      if (!status) {
```
Comment (timshen91) [inline tensorflow/stream_executor/stream.cc]:
```
if (status) {
  status = dnn->DoConvolve(...);
}
if (!status && !output_profile_result) {
  SetError();
}
```

Same for other places.

**Review by timshen91 — COMMENTED** [review]
Thanks for the change Jack! I don't have much to comment about, it's a rather clean patch.

FWIW, admittedly this bloats the interface even more, especially with 10+ arguments per added function. I may attempt to clean them up a bit in the future while not changing the design.

File: tensorflow/stream_executor/dnn.h (hunk 1 above)
Comment (whchung) [reply tensorflow/stream_executor/dnn.h]: done

File: tensorflow/stream_executor/stream.cc (hunk 2 above)
Comment (whchung) [reply tensorflow/stream_executor/stream.cc]: done

**Comment by whchung** [comment]
@timshen91 I've addressed all the issues on CI other than 2 test targets: Windows Bazel, Windows Bazel GPU. From the logs I don't think they are relevant to this PR.

**Review by timshen91 — APPROVED** [review]

**Comment by whchung** [comment]
@timshen91 Just want to understand how to proceed on this PR. I applied `kokoro:force-run` label again and we are stuck on MacOS/Windows builds, but from the logs they have nothing to do with this PR.

BTW, we are about to submit a couple more PRs which includes more refactoring to StreamExecutor which makes larger part it shared between CUDA and ROCm paths.

**Comment by timshen91** [comment]
@jlebar do you know how to get this patch merged? I don't seem to have the option.

**Comment by jlebar** [comment]
> @jlebar do you know how to get this patch merged? I don't seem to have the option.

Forwarded instructions offline, since they're not public.

**Pull request merged**